### PR TITLE
Added a very basic smoke test. Run it with `tox -e smoke`

### PR DESCRIPTION
**Pull request by pengale**
This PR adds a smoke test will build a new "smoke" charm in test/charms/test_smoke.
The test will attempt to deploy xenial, bionic and focal versions of the charm.

## Checklist

 - [N] Have any types changed? If so, have the type annotations been updated?
 - [N/A] If this code exposes model data, does it do so thoughtfully, in a way that aids understanding?

       We want to avoid simple passthrough of model data, without contextualization, where possible.

 - [Y] Do error messages, if any, present charm authors or operators with enough information to solve the problem?

## QA steps

Make sure that you are on a box with charmcraft and juju installed, and that you are connected to a "machine" controller, such as a local lxd cloud. Then run:

```sh
tox -e smoke
```

## Documentation changes

* This adds a step to the release process. Right now, it's a manual step: run `tox -e smoke` before releasing a version of ops.

## Bug reference

* This would have caught https://github.com/canonical/operator/issues/798

## Changelog
- Adds smoke test charm to test/charms dir.
- Adds tox smoke env that builds and deploys the charm.

## Notes
There are a few wonky things about this, including how we inject the current ops package, and the fact that we test with just the focal version of the charm. Future versions of this smoke test should clear those wonky bits up, as well as add automation.

### CODE REVIEW

**Comment by rwcarlsen** [comment]
Would it be worth it to have `tox -e smoke` set up the entire environment inside lxd containers rather than doing stuff directly with juju on my/the host machine?

**Comment by pengale** [comment]
> Would it be worth it to have `tox -e smoke` set up the entire environment inside lxd containers rather than doing stuff directly with juju on my/the host machine?

Eventually, yes!

... though we also want this to work on AWS and other substrates, so you'd need to make it configurable.

But that's out of scope for this PR, which was just "get this thing setup while I'm letting my brain work on dashboard on a background thread."

**Comment by pengale** [comment]
Also, thank you for taking a look :-)

And one more thing: this doesn't work on AWS, because pytest-operator can't handle the self signed cert. I think that's a bug in pytest-operator and/or python-libjuju. I don't think that we need to handle it in this code.

File: test/charms/test_smoke/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# smoke
+
+## Description
+
+A simple test charm for running smoke tests.
+
+## Usage
+
+Run `tox -e smoke` in the root directory of this repository to build and deploy this charm.
```
Comment (rwcarlsen) [inline test/charms/test_smoke/README.md:9]: Could you include the QA steps info from the PR description here - the stuff about how to make sure your environment is prepared to run the tests.  Also probably worth noting that same info as a comment inside tox.ini.

**Review by rwcarlsen — APPROVED** [review]
Looks fine.  Just one small note below about adding some setup explanation in a couple spots.

File: test/smoke/test_smoke.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+
+async def test_smoke(ops_test: OpsTest):
+    # Verify that we can deploy charms from supported series.
+
+    # Build the charm. (We just build it for focal -- it *should* work to deploy it on
+    # older versions of Juju.)
+    charm = await ops_test.build_charm("./test/charms/test_smoke/")
+
+    for series in ['focal', 'bionic', 'xenial']:
```
Comment (PietroPasotti) [inline test/smoke/test_smoke.py:34]: use `pytest.mark.parametrize` to give a better error message (use `ids=['focal', ...]` for bonus points)

File: test/smoke/test_smoke.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+
+import logging
+
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+CURRENT_CHANNEL = '20.04'  # Focal
```
Comment (PietroPasotti) [inline test/smoke/test_smoke.py:24]: where is this used?

File: test/charms/test_smoke/config.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2022 Penelope Valentine Gale
+# See LICENSE file for licensing details.
+#
+# TEMPLATE-TODO: change this example to suit your needs.
+# If you don't need a config, you can remove the file entirely.
+# It ties in to the example _on_config_changed handler in src/charm.py
+#
+# Learn more about config at: https://juju.is/docs/sdk/config
+
+options: {}
```
Comment (PietroPasotti) [inline test/charms/test_smoke/config.yaml:10]: IIRC we can remove the file completely if there's no config

File: test/charms/test_smoke/config.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2022 Penelope Valentine Gale
```
Comment (PietroPasotti) [inline test/charms/test_smoke/config.yaml:1]: Should the copyright not be Canonical?

**Review by PietroPasotti — APPROVED** [review]
Looks good! Just a few minor comments

**Review by rwcarlsen — APPROVED** [review]

File: test/smoke/test_smoke.py (hunk 2 above)
Comment (pengale) [reply test/smoke/test_smoke.py:31]: I started in on this, and it ended up adding a lot of complexity. The build needs to get moved into a separate routine, and then we need a data structure to track our charm artifact. None of that is difficult to write, but it didn't feel worthwhile to do so until this test gets more complex.

File: test/charms/test_smoke/metadata.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2022 Penelope Valentine Gale
+# See LICENSE file for licensing details.
+
```
Comment (rwcarlsen) [inline test/charms/test_smoke/metadata.yaml:3]: Looks like maybe fixing this copyright line got missed before the merge?

**Comment by benhoyt** [comment]
Is this being run by GitHub Actions? If so, I'm wondering why the tests are succeeding when the Xenial fix https://github.com/canonical/operator/pull/810 isn't yet merged?

**Comment by pengale** [comment]
> Is this being run by GitHub Actions? If so, I'm wondering why the tests are succeeding when the Xenial fix #810 isn't yet merged?

The smoke tests are manual tests for now -- this is something that I timeboxed and snuck in as a bonus. 

We intend to automate them in the future. (Probably the next time I need to think about what I'm officially working on for a bit.)

**Pull request merged**